### PR TITLE
Trainer register etcd

### LINKED ISSUE #3051: Register trainer on etcd, so master can sync across all the trainers.

### PR DESCRIPTION
**Pull request by typhoonzero**
Fix #3051

### CODE REVIEW

**Comment by typhoonzero** [comment]
etcd client暂时没有好的方法单测。只能集成测试或者e2e test，如果有好的测试方法，欢迎PR~

File: go/master/etcd_client.go [hunk 1]
```diff
@@ -173,6 +177,44 @@ func GetKey(c *clientv3.Client, key string, timeout time.Duration) (string, erro
...
+		return e
+	}
+	// put trainer uniq id with etcd lease
+	resp, err := c.Grant(context.TODO(), 5)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	_, err = c.Put(ctx, DefaultTrainerPath+"/"+clientUUID.String(), trainerIP, clientv3.WithLease(resp.ID))
```
Comment (Yancey1989) [inline go/master/etcd_client.go]: How about write the pass status into etcd as the following?

```golang
type TrainerStatus struct {
    trainerIP string
    pass_num int
    pass_status string
}
```

For the master check a pass status:
1. master check trainer pass status:
    1.1. if all trainer finished a pass, goto 2
    1.2. else goto 1
2. master move task from Done Queue to Todo Queue.

For the trainer check a pass status:
1. fetch task from master
    1.1. if NoTaskFound, Update status to Finished and goto 2
    1.1. else training with the task
2. if pass_num == finised_pass_num; then finish the train; else goto 1.

**Comment by helinwang** [comment]
@typhoonzero @gongweibao 有了glide之后我们可以用https://godoc.org/github.com/coreos/etcd/embed 了，欢迎PR哈！

**Comment by helinwang** [comment]
Maybe we don't need this PR yet due to the latest change in https://github.com/PaddlePaddle/Paddle/pull/2948 ? If so maybe we can close it, and reopen it once needed :)

**Comment by typhoonzero** [comment]
Closing, reopen when needed.

**Pull request closed without merging**